### ISSUE TITLE
AGS 4: New compiler does not correctly parse integer literals at the negative limit

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#1674]
@fernewelten 

The new compiler has some general mistake in the [`Scanner::ReadInNumberLit`](https://github.com/adventuregamestudio/ags/blob/4a1452a4700535ca7c41c666e6238e46d566ccfa/Compiler/script2/cs_scanner.cpp#L310-L323), where it would incorrectly determine the int32 value range. Possibly this is because it parses sign separately from the numeric value itself, so when testing for overflow it uses the positive value instead.

As a result a valid int32 value `-2147483648` could not be parsed correctly. The code
```
int a = -2147483648;
```
would result in a compiler's error:
`Literal integer '2147483648' is out of bounds (maximum is '2147483647')`
(Notice how it drops the sign in the error message too, and clearly only considers the positive value limit).

This seem to be a know issue, as all the new compiler's tests that test for integer limits use *not* INT32_MIN value, but INT32_MIN + 1 value with comments. For example: https://github.com/adventuregamestudio/ags/blob/4a1452a4700535ca7c41c666e6238e46d566ccfa/Compiler/test2/cc_parser_test_0.cpp#L142
At the same time the scanner test considers valid INT32_MIN to be parsed with error: https://github.com/adventuregamestudio/ags/blob/4a1452a4700535ca7c41c666e6238e46d566ccfa/Compiler/test2/cc_scanner_test.cpp#L322-L329

Another consequence of this mistake in scanner is that it could not correctly parse the large hexadecimal numbers, e.g. `0xFFFFFFFF`, thinking that they exceed the integer limit.

Please note that @AlanDrake made a fix for this latter problem recently (53538101712f81359a4b4162e1142bb4f0c875bb), as it was necessary to let compile the recent additions to the script API (the built-in header). But this fix is not working well, as it created an opposite problem. Now the compiler cannot detect the value overflow correctly. The code
```
    int a = -2147483648;
    int b = -2147483649;
    Display("%d, %d", a, b);
```
will compile, but result in displaying numbers -2147483648 and 2147483647, as internally the variables are still int32.

**Comment by ivan-mogilko** [comment]
Please note that I opened a PR with adjusted tests for this: #1675

**Comment by fernewelten** [comment]
I think that I unwittingly went by the C/C++ semantics without checking what the AGS semantics were. It's not quite trivial to fix because, as Ivan remarked, the scanner reads `- 77'  as `-` `77`, and it's only later in the parser that those two are joined into a negative number.

I'll have a look at this.

**Comment by fernewelten** [comment]
So what we'd get at would be the following specifications:
- `2147483649` and any larger integer is a scanning error.
- `2147483648` is also a scanning error __unless__ preceded by `-`
- Even when the scanner accepts  `- 2147483648`, this might __still__ turn out to be a __parsing__ error, for instance in the expression `i - 2147483648`.

**Comment by AlanDrake** [comment]
Isn't `0x80000000` the one that evaluates to  `-2147483648`?
`0xFFFFFFFF` is `-1`

**Comment by fernewelten** [comment]
@AlanDrake Right you are. Fixed this above.

**Comment by ivan-mogilko** [comment]
> I think that I unwittingly went by the C/C++ semantics without checking what the AGS semantics were.
> it's only later in the parser that those two are joined into a negative number.

@fernewelten I don't know much about this, but imagine this is then a question of Scanner's purpose and what we expect from it. If you say it's meant to scan separate parts of expressions and not know anything about combinations of these elements, than maybe it works correctly in that, but should not error on some situations and let Parser further evaluate them?

**Comment by fernewelten** [comment]
Addressed in #1738.

**Comment by ivan-mogilko** [comment]
Seem to be fixed by #1738
